File: lib/server/workspace_manager.dart
import 'dart:async';
import 'dart:io';
import 'package:logging/logging.dart';
import 'package:path/path.dart' as p;
import '../core/dto.dart';
import '../core/module.dart';
import '../core/workspace.dart';
import 'service_manager.dart';
import 'storage.dart';

final _log = Logger('WorkspaceManager');

class WorkspaceManager {
  /// Where relative `path-on-disk` entries in the .wksp resolve against.
  /// Defaults to the workspace file's parent dir; overridden by the server's
  /// `--clone-root` CLI flag. Absolute paths in the .wksp ignore this.
  final String? overrideCloneRoot;

  String? _workspacePath;
  Workspace? _workspace;
  Map<String, LoadedProjectDto> _projects = {};
  final StreamController<void> _changes = StreamController<void>.broadcast();

  WorkspaceManager({this.overrideCloneRoot});

  String? get workspacePath => _workspacePath;
  Workspace? get workspace => _workspace;
  Map<String, LoadedProjectDto> get projects => Map.unmodifiable(_projects);

  /// Effective root for resolving relative project paths.
  String? get cloneRoot {
    if (overrideCloneRoot != null) return overrideCloneRoot;
    final wp = _workspacePath;
    return wp == null ? null : p.dirname(wp);
  }

  Stream<void> get changes => _changes.stream;

  Future<void> open(String path) async {
    _workspacePath = path;
    _workspace = await Storage.loadWorkspace(path);
    await _reloadProjects();
    _changes.add(null);
  }

  Future<void> close() async {
    _workspacePath = null;
    _workspace = null;
    _projects = {};
    _changes.add(null);
  }

  Future<void> reloadProjects() async {
    if (_workspace == null) return;
    await _reloadProjects();
    _changes.add(null);
  }

  Future<void> _reloadProjects() async {
    final workspacePath = _workspacePath;
    final workspace = _workspace;
    final root = cloneRoot;
    if (workspacePath == null || workspace == null || root == null) {
      _projects = {};
      return;
    }
    final result = <String, LoadedProjectDto>{};
    for (final entry in workspace.projects.entries) {
      result[entry.key] = await _loadOneProject(entry.key, entry.value, root);
    }
    _projects = result;
    _log.info('Loaded ${_projects.length} project(s)');
  }

  Future<LoadedProjectDto> _loadOneProject(
    String name,
    ProjectRef ref,
    String root,
  ) async {
    final resolvedPath = Storage.resolvePath(root, ref.pathOnDisk);
    final exists = await Storage.projectExistsOnDisk(root, ref.pathOnDisk);
    Project? projectData;
    String? loadError;
    if (exists) {
      try {
        projectData = await Storage.loadProject(resolvedPath);
      } on FormatException catch (e) {
        final snippet = Storage.jsonErrorSnippet(
          await File(p.join(resolvedPath, 'limousine.proj')).readAsString(),
          e.offset,
        );
        loadError = 'Invalid JSON in limousine.proj:\n$snippet';
      } catch (e) {
        loadError = 'Failed to load project: $e';
      }
    }
    return LoadedProjectDto(
      name: name,
      resolvedPath: resolvedPath,
      gitRepoUrl: ref.gitRepoUrl,
      existsOnDisk: exists,
      projectData: projectData,
      loadError: loadError,
    );
  }

  /// Reload one project's `limousine.proj` without touching the rest of the
  /// workspace. Caller is responsible for checking that no services in the
  /// project are running.
  Future<void> reloadProject(String name) async {
    final workspace = _workspace;
    final root = cloneRoot;
    if (workspace == null || root == null) return;
    final ref = workspace.projects[name];
    if (ref == null) return;
    _projects = {
      ..._projects,
      name: await _loadOneProject(name, ref, root),
    };
    _changes.add(null);
    _log.info('Reloaded project: $name');
  }

  /// Re-read the `.wksp` from disk and compute the diff against the current
  /// in-memory state. Doesn't apply anything — caller decides whether the
  /// diff is safe given current service state, then calls [applyDiff].
  Future<WorkspaceDiff> peekDiff() async {
    final path = _workspacePath;
    final current = _workspace;
    if (path == null || current == null) {
      throw StateError('No workspace open');
    }
    final next = await Storage.loadWorkspace(path);
    final added = <String>[];
    final removed = <String>[];
    final changed = <String>[];
    for (final entry in next.projects.entries) {
      final old = current.projects[entry.key];
      if (old == null) {
        added.add(entry.key);
      } else if (old.pathOnDisk != entry.value.pathOnDisk ||
          old.gitRepoUrl != entry.value.gitRepoUrl) {
        changed.add(entry.key);
      }
    }
    for (final oldKey in current.projects.keys) {
      if (!next.projects.containsKey(oldKey)) removed.add(oldKey);
    }
    return WorkspaceDiff(
      added: added,
      removed: removed,
      changed: changed,
      newWorkspace: next,
    );
  }

  /// Apply a previously-peeked diff. Removed projects are unloaded; added and
  /// changed projects are (re)loaded against the new ProjectRef.
  Future<void> applyDiff(WorkspaceDiff diff) async {
    final root = cloneRoot;
    if (root == null) return;
    _workspace = diff.newWorkspace;
    final next = {..._projects};
    for (final name in diff.removed) {
      next.remove(name);
    }
    for (final name in [...diff.added, ...diff.changed]) {
      final ref = _workspace!.projects[name];
      if (ref != null) next[name] = await _loadOneProject(name, ref, root);
    }
    _projects = next;
    _changes.add(null);
    _log.info(
      'Workspace reload applied: +${diff.added.length} -${diff.removed.length} ~${diff.changed.length}',
    );
  }

  List<ServiceInfo> allServices() {
    final result = <ServiceInfo>[];
    for (final project in _projects.values) {
      if (project.projectData == null) continue;
      for (final module in project.projectData!.modules) {
        for (final entry in module.services.entries) {
          result.add(ServiceInfo(
            projectName: project.name,
            projectPath: project.resolvedPath,
            moduleName: module.name,
            moduleConfig: module.config,
            serviceName: entry.key,
            service: entry.value,
          ));
        }
      }
    }
    return result;
  }

  ServiceInfo? findService(String serviceId) {
    return allServices().where((s) => s.id == serviceId).firstOrNull;
  }

  Future<void> saveWorkspace(Workspace updated) async {
    final path = _workspacePath;
    if (path == null) return;
    await Storage.saveWorkspace(path, updated);
    _workspace = updated;
    _changes.add(null);
  }

  Future<void> dispose() async {
    await _changes.close();
  }
}

class WorkspaceDiff {
  final List<String> added;
  final List<String> removed;
  final List<String> changed;
  final Workspace newWorkspace;

  WorkspaceDiff({
    required this.added,
    required this.removed,
    required this.changed,
    required this.newWorkspace,
  });

  bool get hasChanges =>
      added.isNotEmpty || removed.isNotEmpty || changed.isNotEmpty;
}

File: limousine/backend/workspace_manager.py
"""In-memory state for the open workspace + loaded projects. Port of
lib/server/workspace_manager.dart. Change notification is a simple callback list
(the Dart broadcast stream); the TUI/backend subscribe to refresh on edits."""

from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path
from typing import Callable

from ..core.dtos import LoadedProject
from ..core.models import ProjectRef, Workspace
from . import storage
from .manager import ServiceInfo


@dataclass
class WorkspaceDiff:
    added: list[str]
    removed: list[str]
    changed: list[str]
    new_workspace: Workspace

    @property
    def has_changes(self) -> bool:
        return bool(self.added or self.removed or self.changed)


class WorkspaceManager:
    def __init__(self, override_clone_root: str | None = None):
        self.override_clone_root = override_clone_root
        self._workspace_path: str | None = None
        self._workspace: Workspace | None = None
        self._projects: dict[str, LoadedProject] = {}
        self._listeners: list[Callable[[], None]] = []

    # ---- accessors -------------------------------------------------------

    @property
    def workspace_path(self) -> str | None:
        return self._workspace_path

    @property
    def workspace(self) -> Workspace | None:
        return self._workspace

    @property
    def projects(self) -> dict[str, LoadedProject]:
        return dict(self._projects)

    @property
    def clone_root(self) -> str | None:
        if self.override_clone_root is not None:
            return self.override_clone_root
        return str(Path(self._workspace_path).parent) if self._workspace_path else None

    def add_listener(self, cb: Callable[[], None]) -> None:
        self._listeners.append(cb)

    def _notify(self) -> None:
        for cb in self._listeners:
            cb()

    # ---- open / close / reload ------------------------------------------

    def open(self, path: str) -> None:
        self._workspace_path = path
        self._workspace = storage.load_workspace(path)
        self._reload_projects()
        self._notify()

    def close(self) -> None:
        self._workspace_path = None
        self._workspace = None
        self._projects = {}
        self._notify()

    def reload_projects(self) -> None:
        if self._workspace is None:
            return
        self._reload_projects()
        self._notify()

    def _reload_projects(self) -> None:
        root = self.clone_root
        if self._workspace is None or root is None:
            self._projects = {}
            return
        self._projects = {
            name: self._load_one(name, ref, root) for name, ref in self._workspace.projects.items()
        }

    def _load_one(self, name: str, ref: ProjectRef, root: str) -> LoadedProject:
        resolved = storage.resolve_path(root, ref.path_on_disk)
        exists = storage.project_exists_on_disk(root, ref.path_on_disk)
        project_data = None
        load_error = None
        if exists:
            try:
                project_data = storage.load_project(resolved)
            except ValueError as e:
                load_error = str(e)
            except Exception as e:  # surface dev-only failures with detail
                load_error = f"Failed to load project: {e}"
        return LoadedProject(
            name=name,
            resolved_path=resolved,
            git_repo_url=ref.git_repo_url,
            exists_on_disk=exists,
            project_data=project_data,
            load_error=load_error,
        )

    def reload_project(self, name: str) -> None:
        root = self.clone_root
        if self._workspace is None or root is None:
            return
        ref = self._workspace.projects.get(name)
        if ref is None:
            return
        self._projects[name] = self._load_one(name, ref, root)
        self._notify()

    # ---- diff-based workspace reload ------------------------------------

    def peek_diff(self) -> WorkspaceDiff:
        if self._workspace_path is None or self._workspace is None:
            raise RuntimeError("No workspace open")
        nxt = storage.load_workspace(self._workspace_path)
        added, removed, changed = [], [], []
        for key, ref in nxt.projects.items():
            old = self._workspace.projects.get(key)
            if old is None:
                added.append(key)
            elif old.path_on_disk != ref.path_on_disk or old.git_repo_url != ref.git_repo_url:
                changed.append(key)
        for old_key in self._workspace.projects:
            if old_key not in nxt.projects:
                removed.append(old_key)
        return WorkspaceDiff(added=added, removed=removed, changed=changed, new_workspace=nxt)

    def apply_diff(self, diff: WorkspaceDiff) -> None:
        root = self.clone_root
        if root is None:
            return
        self._workspace = diff.new_workspace
        for name in diff.removed:
            self._projects.pop(name, None)
        for name in [*diff.added, *diff.changed]:
            ref = self._workspace.projects.get(name)
            if ref is not None:
                self._projects[name] = self._load_one(name, ref, root)
        self._notify()

    # ---- services --------------------------------------------------------

    def all_services(self) -> list[ServiceInfo]:
        result: list[ServiceInfo] = []
        for project in self._projects.values():
            if project.project_data is None:
                continue
            for module in project.project_data.modules:
                for svc_name, service in module.services.items():
                    result.append(
                        ServiceInfo(
                            project_name=project.name,
                            project_path=project.resolved_path,
                            module_name=module.name,
                            module_config=module.config,
                            service_name=svc_name,
                            service=service,
                        )
                    )
        return result

    def find_service(self, service_id: str) -> ServiceInfo | None:
        return next((s for s in self.all_services() if s.id == service_id), None)

    def save_workspace(self, updated: Workspace) -> None:
        if self._workspace_path is None:
            return
        storage.save_workspace(self._workspace_path, updated)
        self._workspace = updated
        self._notify()
